File: src/geom/src/stl.cpp
// SPDX-License-Identifier: BSD-3-Clause
#include "geom/stl.hpp"

#include <bit>
#include <cstring>
#include <format>
#include <fstream>
#include <map>
#include <sstream>

namespace polymesh::geom {
namespace detail {
namespace {

// STL files are little-endian; supporting big-endian hosts would need
// byte-swapping here.
static_assert(std::endian::native == std::endian::little);

/// Reads a little-endian scalar at byte offset `at`.
template <typename T> T read_le(std::span<const std::byte> bytes, std::size_t at) {
    T value;
    std::memcpy(&value, bytes.data() + at, sizeof(T));
    return value;
}

bool is_ascii_stl(std::span<const std::byte> bytes) {
    // A binary STL can also start with "solid"; require an ASCII "facet"
    // token near the top to disambiguate.
    constexpr std::string_view kSolid = "solid";
    if (bytes.size() < kSolid.size() ||
        std::memcmp(bytes.data(), kSolid.data(), kSolid.size()) != 0) {
        return false;
    }
    const std::string_view head(reinterpret_cast<const char*>(bytes.data()),
                                std::min<std::size_t>(bytes.size(), 1024));
    return head.find("facet") != std::string_view::npos;
}

} // namespace

Soup parse_binary(std::span<const std::byte> bytes) {
    constexpr std::size_t kHeader = 84;
    constexpr std::size_t kRecord = 50;
    if (bytes.size() < kHeader) {
        throw GeomError("malformed STL: shorter than 84-byte header");
    }
    const auto n = read_le<std::uint32_t>(bytes, 80);
    const std::size_t expected = kHeader + std::size_t{n} * kRecord;
    if (bytes.size() < expected) {
        throw GeomError(std::format(
            "malformed STL: header declares {} triangles ({} bytes) but file has {}", n,
            expected, bytes.size()));
    }
    Soup soup;
    soup.reserve(n);
    for (std::size_t t = 0; t < n; ++t) {
        // Skip the 12-byte stored normal; it is recomputed from vertices.
        const std::size_t base = kHeader + t * kRecord + 12;
        std::array<double, 9> tri{};
        for (std::size_t k = 0; k < 9; ++k) {
            tri[k] = static_cast<double>(read_le<float>(bytes, base + k * 4));
        }
        soup.push_back(tri);
    }
    return soup;
}

Soup parse_ascii(std::span<const std::byte> bytes) {
    const std::string_view text(reinterpret_cast<const char*>(bytes.data()), bytes.size());
    Soup soup;
    std::array<double, 9> coords{};
    std::size_t filled = 0;
    std::istringstream lines{std::string(text)};
    std::string line;
    while (std::getline(lines, line)) {
        std::istringstream words(line);
        std::string keyword;
        words >> keyword;
        if (keyword != "vertex") {
            continue;
        }
        for (int k = 0; k < 3; ++k) {
            double value = 0.0;
            if (!(words >> value)) {
                throw GeomError("malformed STL: vertex with fewer than 3 coordinates");
            }
            coords[filled++] = value;
        }
        if (filled == 9) {
            soup.push_back(coords);
            filled = 0;
        }
    }
    if (filled != 0) {
        throw GeomError("malformed STL: facet with incomplete vertex triple");
    }
    return soup;
}

TriSurface weld(const Soup& soup) {
    // Key on the raw bit patterns for exact, tolerance-free deduplication.
    std::map<std::array<std::uint64_t, 3>, std::uint32_t> seen;
    TriSurface surface;
    for (const auto& tri : soup) {
        std::array<std::uint32_t, 3> idx{};
        for (std::size_t v = 0; v < 3; ++v) {
            const std::array<double, 3> p{tri[v * 3], tri[v * 3 + 1], tri[v * 3 + 2]};
            const std::array<std::uint64_t, 3> key{std::bit_cast<std::uint64_t>(p[0]),
                                                   std::bit_cast<std::uint64_t>(p[1]),
                                                   std::bit_cast<std::uint64_t>(p[2])};
            const auto [it, inserted] =
                seen.try_emplace(key, static_cast<std::uint32_t>(surface.vertices.size()));
            if (inserted) {
                surface.vertices.emplace_back(p[0], p[1], p[2]);
            }
            idx[v] = it->second;
        }
        surface.triangles.push_back(idx);
    }
    return surface;
}

} // namespace detail

TriSurface load_stl(const std::filesystem::path& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw GeomError(std::format("cannot open {}", path.string()));
    }
    std::vector<std::byte> bytes(std::filesystem::file_size(path));
    file.read(reinterpret_cast<char*>(bytes.data()),
              static_cast<std::streamsize>(bytes.size()));
    if (!file) {
        throw GeomError(std::format("i/o error reading {}", path.string()));
    }
    const auto soup =
        detail::is_ascii_stl(bytes) ? detail::parse_ascii(bytes) : detail::parse_binary(bytes);
    return detail::weld(soup);
}

} // namespace polymesh::geom

File: crates/geom/src/stl.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Binary and ASCII STL loading with vertex welding.

use crate::{GeomError, TriSurface};
use nalgebra::Point3;
use std::collections::HashMap;
use std::io::Read;
use std::path::Path;

/// Loads an STL file (binary or ASCII, auto-detected) and welds duplicate
/// vertices exactly (bitwise equality — STL repeats vertices verbatim, so
/// exact welding is deterministic and tolerance-free).
pub fn load_stl(path: &Path) -> Result<TriSurface, GeomError> {
    let mut bytes = Vec::new();
    std::fs::File::open(path)?.read_to_end(&mut bytes)?;
    let raw = if is_ascii_stl(&bytes) {
        parse_ascii(&bytes)?
    } else {
        parse_binary(&bytes)?
    };
    Ok(weld(&raw))
}

/// Raw triangle soup: 9 coordinates per triangle.
type Soup = Vec<[f64; 9]>;

fn is_ascii_stl(bytes: &[u8]) -> bool {
    // A binary STL can also start with "solid"; require an ASCII "facet"
    // token in the body to disambiguate.
    bytes.starts_with(b"solid")
        && std::str::from_utf8(&bytes[..bytes.len().min(1024)])
            .map(|s| s.contains("facet"))
            .unwrap_or(false)
}

fn parse_binary(bytes: &[u8]) -> Result<Soup, GeomError> {
    if bytes.len() < 84 {
        return Err(GeomError::MalformedStl(
            "shorter than 84-byte header".into(),
        ));
    }
    let n = u32::from_le_bytes(bytes[80..84].try_into().unwrap()) as usize;
    let expected = 84 + n * 50;
    if bytes.len() < expected {
        return Err(GeomError::MalformedStl(format!(
            "header declares {n} triangles ({expected} bytes) but file has {}",
            bytes.len()
        )));
    }
    let mut soup = Vec::with_capacity(n);
    for t in 0..n {
        let base = 84 + t * 50 + 12; // skip the normal, recompute from vertices
        let mut tri = [0.0f64; 9];
        for (k, v) in tri.iter_mut().enumerate() {
            let o = base + k * 4;
            *v = f32::from_le_bytes(bytes[o..o + 4].try_into().unwrap()) as f64;
        }
        soup.push(tri);
    }
    Ok(soup)
}

fn parse_ascii(bytes: &[u8]) -> Result<Soup, GeomError> {
    let text = std::str::from_utf8(bytes)
        .map_err(|e| GeomError::MalformedStl(format!("invalid utf-8: {e}")))?;
    let mut soup = Vec::new();
    let mut coords: Vec<f64> = Vec::with_capacity(9);
    for line in text.lines() {
        let mut it = line.split_whitespace();
        if it.next() == Some("vertex") {
            for _ in 0..3 {
                let tok = it
                    .next()
                    .ok_or_else(|| GeomError::MalformedStl("vertex with <3 coords".into()))?;
                coords.push(
                    tok.parse()
                        .map_err(|e| GeomError::MalformedStl(format!("bad float {tok:?}: {e}")))?,
                );
            }
            if coords.len() == 9 {
                soup.push(coords[..].try_into().unwrap());
                coords.clear();
            }
        }
    }
    if !coords.is_empty() {
        return Err(GeomError::MalformedStl(
            "facet with incomplete vertex triple".into(),
        ));
    }
    Ok(soup)
}

fn weld(soup: &Soup) -> TriSurface {
    let mut map: HashMap<[u64; 3], u32> = HashMap::new();
    let mut surface = TriSurface::default();
    for tri in soup {
        let mut idx = [0u32; 3];
        for (v, slot) in idx.iter_mut().enumerate() {
            let p = [tri[v * 3], tri[v * 3 + 1], tri[v * 3 + 2]];
            let key = p.map(f64::to_bits);
            *slot = *map.entry(key).or_insert_with(|| {
                surface.vertices.push(Point3::from(p));
                (surface.vertices.len() - 1) as u32
            });
        }
        surface.triangles.push(idx);
    }
    surface
}

#[cfg(test)]
mod tests {
    use super::*;

    const TETRA_ASCII: &str = "\
solid tetra
 facet normal 0 0 -1
  outer loop
   vertex 0 0 0
   vertex 0 1 0
   vertex 1 0 0
  endloop
 endfacet
 facet normal 0 -1 0
  outer loop
   vertex 0 0 0
   vertex 1 0 0
   vertex 0 0 1
  endloop
 endfacet
 facet normal -1 0 0
  outer loop
   vertex 0 0 0
   vertex 0 0 1
   vertex 0 1 0
  endloop
 endfacet
 facet normal 1 1 1
  outer loop
   vertex 1 0 0
   vertex 0 1 0
   vertex 0 0 1
  endloop
 endfacet
endsolid tetra
";

    #[test]
    fn ascii_tetra_welds_to_four_vertices() {
        let soup = parse_ascii(TETRA_ASCII.as_bytes()).unwrap();
        assert_eq!(soup.len(), 4);
        let s = weld(&soup);
        assert_eq!(s.vertices.len(), 4);
        assert_eq!(s.triangles.len(), 4);
        s.validate().unwrap();
    }

    #[test]
    fn binary_roundtrip() {
        // Build a binary STL for one triangle and parse it back.
        let mut bytes = vec![0u8; 84];
        bytes[80..84].copy_from_slice(&1u32.to_le_bytes());
        let mut rec = vec![0u8; 12]; // normal
        for v in [[0f32, 0., 0.], [1., 0., 0.], [0., 1., 0.]] {
            for c in v {
                rec.extend_from_slice(&c.to_le_bytes());
            }
        }
        rec.extend_from_slice(&[0, 0]); // attribute byte count
        bytes.extend_from_slice(&rec);
        let soup = parse_binary(&bytes).unwrap();
        assert_eq!(soup.len(), 1);
        assert_eq!(soup[0][3], 1.0);
    }
}
